### PR TITLE
Do not hardcode the media manager API url to select an image

### PR DESCRIPTION
**Pull request by laoneo**
### Summary of Changes
The media manager is loading their API urls from the JS config store, expect when selecting files in the media form field. This pr extends the media manager the way that the API url is passed to the media field. For backwards compatibility it does a fallback to the old url when it doesn't exist already. This situation can be for older overrides.

@dgrammatiko can you review?

### Testing Instructions
Select an intro or full image in the images field.

### Actual result BEFORE applying this Pull Request
All works.

### Expected result AFTER applying this Pull Request
All works.

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [x] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [x] No documentation changes for manual.joomla.org needed

### CODE REVIEW

File: build/media_source/system/js/fields/joomla-media-select.w-c.es6.js [hunk 1]
```diff
@@ -317,7 +317,8 @@ Joomla.getMedia = (data, editor, fieldClass) => new Promise((resolve, reject) =>
     return;
   }
 
-  const url = `${Joomla.getOptions('system.paths').baseFull}index.php?option=com_media&task=api.files&url=true&path=${encodeURIComponent(data.path)}&mediatypes=0,1,2,3&${Joomla.getOptions('csrf.token')}=1&format=json`;
+  let url = `&task=api.files&url=true&path=${encodeURIComponent(data.path)}&mediatypes=0,1,2,3&${Joomla.getOptions('csrf.token')}=1`;
+  url = (Joomla.getOptions('media-picker-api').apiBaseUrl ? Joomla.getOptions('media-picker-api').apiBaseUrl : Joomla.getOptions('system.paths').baseFull +'index.php?option=com_media&format=json') + url;
```
Comment (dgrammatiko) [inline build/media_source/system/js/fields/joomla-media-select.w-c.es6.js]: May I suggest replacing these 2 lines with:
```js
const url = new URL(Joomla.getOptions('media-picker-api').apiBaseUrl ? Joomla.getOptions('media-picker-api').apiBaseUrl : Joomla.getOptions('system.paths').baseFull +'index.php?option=com_media&format=json');
url.searchParams.append('task', 'api.files');
url.searchParams.append('url', true);
url.searchParams.append('path', encodeURIComponent(data.path));
url.searchParams.append('mediatypes', '0,1,2,3');
url.searchParams.append(Joomla.getOptions('csrf.token'), 1);
```

Comment (laoneo) [reply build/media_source/system/js/fields/joomla-media-select.w-c.es6.js]: Thanks, added the code, looks much nicer

**Comment by dgrammatiko** [comment]
I have tested this item :white_check_mark: successfully on 2fe8e36e3ae2d24de5a92eabb1fae0ecfbe36517<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39647">issues.joomla.org/tracker/joomla-cms/39647</a>.</sub>

**Comment by dgrammatiko** [comment]
Out of curiosity, why did you went with a new entry in the storageOption instead of reusing the existing `media-picker` key?
Screenshot 2023-01-16 at 17 55 52

**Comment by richard67** [comment]
@laoneo The JavaScript linter complains: https://ci.joomla.org/joomla/joomla-cms/60937/1/29

**Comment by laoneo** [comment]
> Out of curiosity, why did you went with a new entry in the storageOption instead of reusing the existing `media-picker` key? Screenshot 2023-01-16 at 17 55 52

Because the old code which adds the entry makes a check if it doesn't exist and then it adds it otherwise not. Using that one makes it unnecessary complex.

**Comment by dgrammatiko** [comment]
Makes sense, thanks

**Comment by carlitorweb** [comment]
I have tested this item :white_check_mark: successfully on 7cdd1a5ac92853237204f221aba4f59d381d2302<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39647">issues.joomla.org/tracker/joomla-cms/39647</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39647">issues.joomla.org/tracker/joomla-cms/39647</a>.</sub>

**Comment by carlitorweb** [comment]
@dgrammatiko I just see with this PR the TODO comment: `// @TODO add this attribute to the field in order to use it for all media types` in the layout file. This is related with the solution of the issue #35088 ?

**Comment by dgrammatiko** [comment]
Yup, that's the missing part

**Comment by carlitorweb** [comment]
I wonder if is better insert the types allowed from the options in com_media, instead of hardcode it like this.

File: build/media_source/system/js/fields/joomla-media-select.w-c.es6.js [hunk 2]
```diff
@@ -317,7 +317,14 @@ Joomla.getMedia = (data, editor, fieldClass) => new Promise((resolve, reject) =>
     return;
   }
 
-  const url = `${Joomla.getOptions('system.paths').baseFull}index.php?option=com_media&task=api.files&url=true&path=${encodeURIComponent(data.path)}&mediatypes=0,1,2,3&${Joomla.getOptions('csrf.token')}=1&format=json`;
+  // Compile the url
+  const url = new URL(Joomla.getOptions('media-picker-api').apiBaseUrl ? Joomla.getOptions('media-picker-api').apiBaseUrl : Joomla.getOptions('system.paths').baseFull + 'index.php?option=com_media&format=json');
```
Comment (dgrammatiko) [inline build/media_source/system/js/fields/joomla-media-select.w-c.es6.js]:
```suggestion
  const url = new URL(Joomla.getOptions('media-picker-api').apiBaseUrl ? Joomla.getOptions('media-picker-api').apiBaseUrl : `${Joomla.getOptions('system.paths').baseFull}index.php?option=com_media&format=json`);
```

Should fix CS

File: build/media_source/system/js/fields/joomla-media-select.w-c.es6.js [hunk 3]
```diff
@@ -317,7 +317,14 @@ Joomla.getMedia = (data, editor, fieldClass) => new Promise((resolve, reject) =>
...
+  url.searchParams.append('task', 'api.files');
+  url.searchParams.append('url', true);
+  url.searchParams.append('path', data.path);
```
Comment (dgrammatiko) [inline build/media_source/system/js/fields/joomla-media-select.w-c.es6.js:324]:
```suggestion
  url.searchParams.append('path', encodeURIComponent(data.path));
```

Please keep the `encodeURIComponent` as it seems that the browsers don't do that automatically: https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams#preserving_plus_signs

**Comment by dgrammatiko** [comment]
I have tested this item :white_check_mark: successfully on aecc96fad1c8b831340fa4bbc9ea93c407e3e22f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39647">issues.joomla.org/tracker/joomla-cms/39647</a>.</sub>

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on aecc96fad1c8b831340fa4bbc9ea93c407e3e22f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39647">issues.joomla.org/tracker/joomla-cms/39647</a>.</sub>

**Comment by dgrammatiko** [comment]
> I wonder if is better insert the types allowed from the options in com_media, instead of hardcode it like this.

@carlitorweb I'm not sure if the MediaHelper returns the list needed here (probably it does because I think I used it in the core media field). Also the implementation should be fairly easy, remove the hardcoded `types`, add a subform field list for the `types` (default value to images for b/c). The only thing that might need some thinking is the layout (to keep it 100% b/c), my personal idea would be to keep the default layout for images and add another when there are more options selected (the list field should be multiple, meaning the value could be any combination between `images`, `audios`, `videos`, `documents`).

**Comment by dgrammatiko** [comment]
I have tested this item :white_check_mark: successfully on 656f9aa4ab97f3d38478389e160d82f31b218eae<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39647">issues.joomla.org/tracker/joomla-cms/39647</a>.</sub>

**Comment by obuisard** [comment]
Thank you Allon @laoneo !

**Pull request merged**